File: backend/app/routes/tasks.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from app.services.task_service import TaskService

bp = Blueprint('tasks', __name__)


@bp.route('', methods=['POST'])
@jwt_required()
def create_task():
    """Создать новую задачу"""
    current_user_id = get_jwt_identity()
    data = request.get_json()
    
    if not data:
        return jsonify({'error': 'No data provided'}), 400
    
    task, error = TaskService.create_task(data, current_user_id)
    
    if error:
        return jsonify({'error': error}), 400
    
    return jsonify({
        'message': 'Task created successfully',
        'task': task.to_dict()
    }), 201


@bp.route('/project/<int:project_id>', methods=['GET'])
@jwt_required()
def get_project_tasks(project_id):
    """Получить задачи проекта"""
    current_user_id = get_jwt_identity()
    status_filter = request.args.get('status')
    
    tasks, error = TaskService.get_tasks(project_id, current_user_id, status_filter)
    
    if error:
        return jsonify({'error': error}), 400
    
    return jsonify({
        'tasks': [t.to_dict(include_raci=True) for t in tasks]
    }), 200


@bp.route('/<int:task_id>', methods=['GET'])
@jwt_required()
def get_task(task_id):
    """Получить задачу по ID"""
    current_user_id = get_jwt_identity()
    
    task, error = TaskService.get_task(task_id, current_user_id)
    
    if error:
        return jsonify({'error': error}), 404
    
    return jsonify({
        'task': task.to_dict(include_raci=True)
    }), 200


@bp.route('/<int:task_id>', methods=['PUT'])
@jwt_required()
def update_task(task_id):
    """Обновить задачу"""
    current_user_id = get_jwt_identity()
    data = request.get_json()
    
    if not data:
        return jsonify({'error': 'No data provided'}), 400
    
    task, error = TaskService.update_task(task_id, data, current_user_id)
    
    if error:
        return jsonify({'error': error}), 400
    
    return jsonify({
        'message': 'Task updated successfully',
        'task': task.to_dict(include_raci=True)
    }), 200


@bp.route('/<int:task_id>', methods=['DELETE'])
@jwt_required()
def delete_task(task_id):
    """Удалить задачу"""
    current_user_id = get_jwt_identity()
    
    result, error = TaskService.delete_task(task_id, current_user_id)
    
    if error:
        return jsonify({'error': error}), 400
    
    return jsonify({
        'message': 'Task deleted successfully'
    }), 200


@bp.route('/my-tasks', methods=['GET'])
@jwt_required()
def get_my_tasks():
    """Получить мои задачи"""
    current_user_id = get_jwt_identity()
    role_filter = request.args.get('role', 'RESPONSIBLE')
    
    tasks, error = TaskService.get_user_tasks(current_user_id, role_filter)
    
    if error:
        return jsonify({'error': error}), 400
    
    return jsonify({
        'tasks': [t.to_dict(include_raci=True) for t in tasks]
    }), 200

File: frontend/src/api/tasks.ts
import { api } from "./axios";
import { Task, RACIRole } from "@/types";

export const tasksApi = {
  // Получить мои задачи
  getMyTasks: async (role?: RACIRole): Promise<Task[]> => {
    const params = role ? { role } : {};
    const response = await api.get("/tasks/my-tasks", { params });
    
    if (Array.isArray(response.data)) {
      return response.data;
    }
    
    if (response.data.tasks && Array.isArray(response.data.tasks)) {
      return response.data.tasks;
    }
    
    if (response.data.data && Array.isArray(response.data.data)) {
      return response.data.data;
    }
    
    return [];
  },

  // Получить задачи проекта
  getProjectTasks: async (projectId: number, status?: string): Promise<Task[]> => {
    const params = status ? { status } : {};
    const response = await api.get(`/tasks/project/${projectId}`, { params });
    
    if (Array.isArray(response.data)) {
      return response.data;
    }
    
    if (response.data.tasks && Array.isArray(response.data.tasks)) {
      return response.data.tasks;
    }
    
    return [];
  },

  // Получить задачу по ID
  getTask: async (taskId: number): Promise<Task> => {
    const response = await api.get(`/tasks/${taskId}`);
    return response.data.task || response.data;
  },

  // Создать задачу
  createTask: async (data: {
    project_id: number;
    title: string;
    description: string;
    priority: number;
    deadline?: string;
  }): Promise<Task> => {
    const response = await api.post("/tasks", data);
    return response.data.task || response.data;
  },

  // Обновить задачу
  updateTask: async (taskId: number, data: Partial<Task>): Promise<Task> => {
    const response = await api.put(`/tasks/${taskId}`, data);
    return response.data.task || response.data;
  },

  // Удалить задачу
  deleteTask: async (taskId: number): Promise<void> => {
    await api.delete(`/tasks/${taskId}`);
  },
};
